### PR TITLE
Pass slsa-related config via explicit struct instead of separate args

### PR DESCRIPTION
**Pull request by chuangw6**
# Changes

/cleanup

Prior, we received slsa-related configs from configmap and passed them
via separate arguments down to GenerateAttestation.

Now, we pass them via a struct as a whole instead of separate args. This
will enable slsa formatters to be extensible to receive more slsa-related
configs without having to change downstream functions' interface i.e.
config option [`ObserveMode`](https://github.com/tektoncd/chains/pull/866#discussion_r1269715883)
which customizes how Chains observes inputs/outputs.

Signed-off-by: Chuang Wang <chuangw@google.com>
@chuangw6

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [x] Has [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)

# Release Notes

``` release-note
NONE
```

### CODE REVIEW

**Comment by chuangw6** [comment]
Note: this change currently only applies to slsa v1 provenance. Will make changes to other slsa formatters after collecting some feedback on this approach.

**Comment by chuangw6** [comment]
Looking for feedback and inputs. Thanks


cc @chitrangpatel @lcarva @wlynch

**Comment by chitrangpatel** [comment]
This is great! I think we should also implement this for other formats `v1 & v2alpha1`.

**Review by wlynch — DISMISSED** [review]
I would recommend against doing this. Context values are tricky because it becomes less-obvious that certain values need to be set.

My general rule of thumb is that callers should be able to operate normally if context values are not present, but here I think BuilderID is critical.

I'd recommend taking a look at the v1 interface proposal, particularly [Payloader](https://github.com/tektoncd/chains/blob/main/docs/v1-proposal.md#payloaders). I think this will do what you want, but with an explicit struct field instead.

**Comment by chuangw6** [comment]
Thank you for the feedback on the context! Concur that context is a less good solution.

> > be extensible to receive more slsa-related configs without having to change downstream functions' interface i.e. config option https://github.com/tektoncd/chains/pull/866#discussion_r1269715883 which customizes how Chains observes inputs/outputs.
>
> An alternative to achieve that benefit is to make GenerateAttestation functions take a simple struct, something like a SLSAConfig.

Yeah, this is the alternative that I thought before other than context -  passing the explicit struct. (worth noting that unlike context, if we need to pass down the configs beyond GenerateAttestation i.e. to subjects / materials, we need to change their interface to accept the struct as well. But I think it's fine given that context is not best solution.) 

Just changed the code to the struct solution. PTAL.


> I'd recommend taking a look at the v1 interface proposal, particularly [Payloader](https://github.com/tektoncd/chains/blob/main/docs/v1-proposal.md#payloaders). I think this will do what you want, but with an explicit struct field instead.

@wlynch is the solution mentioned above what you mean here?

**Comment by chuangw6** [comment]
The new change is ready for review. PTAL @chitrangpatel @wlynch.

Thanks!

**Comment by chitrangpatel** [comment]
/approve

I'll let @wlynch lgtm since he had comments regarding the original PR.

**Comment by wlynch** [comment]
> @wlynch is the solution mentioned above what you mean here?

No, but it's probably going to change soon so 🤷

**Comment by chuangw6** [comment]
Looks like the PR merge is blocked by the requested changes. Maybe need approval as well? @wlynch

**Pull request merged**
